### ISSUE TITLE
Should we have a custom Prelude for tests?

### ISSUE THREAD

**Issue by rbasso** [exercism/haskell#570]
Sometimes the standard `Prelude` makes writing beautiful code a really painful experience, forcing us to import a lot of stuff even to write basic code. Programmers tend to avoid imports, and this favors some bad coding practices that use functions already in the standard Prelude.

Most of the times, the imports we have depend only on `base`...

##### Package dependencies

```plain
 1 array      (saddle-points)
 1 criterion  (parallel-letter-frequency)
 1 multiset   (go-counting)
 1 vector     (matrix)
 3 QuickCheck (trinary, simple-linked-list, octal)
 3 text       (parallel-letter-frequency, forth, sgf-parsing)
 3 time       (lens-person, gigasecond, meetup)
 6 containers (parallel-letter-frequency, go-counting, etl, nucleotide-count, sgf-parsing, pov)
82 base       all
82 hspec      all
```

###### Available in `BasePrelude`, from `base-prelude` package

```haskell
Control.Arrow      ((&&&))
Control.Concurrent (forkIO, newEmptyMVar, putMVar, takeMVar)
Control.Exception  (Exception, throw, evaluate)
Control.Monad      (foldM, forM_, replicateM, unless)
Data.Bifunctor     (bimap, first)
Data.Char          (intToDigit, isSpace, toLower, toUpper)
Data.Either        (isLeft)
Data.Foldable      (for_)
Data.Function      (on)
Data.Ix            (inRange)
Data.Maybe         (fromJust)
Data.Tuple         (swap)
Data.List          (isPrefixOf, nub, sort)
Text.Printf        (printf)
Numeric            (showIntAtBase, showOct)
```

###### Not in `BasePrelude`

GHC.Exts (toList)

##### Imports from other packages

```haskell
array       Data.Array                (listArray)
containers  Data.Map                  (empty, fromList, lookup, singleton)
containers  Data.Set                  (toAscList)
containers  Data.Tree                 (Tree(Node), rootLabel)
multiset    Data.MultiSet             (fromOccurList, toOccurList)
text        Data.Text                 (concat)
time        Data.Time.Calendar        (fromGregorian)
time        Data.Time.Clock           (UTCTime)
time        Data.Time.Format          (ParseTime, TimeLocale, defaultTimeLocale,
                                       iso8601DateFormat, parseTimeOrError)
vector      Data.Vector               (fromList)
hspec       Test.Hspec.Runner         (configFastFail, defaultConfig, hspecWith)
hspec       Test.Hspec                (Spec, describe, expectationFailure, it,
                                       shouldBe, shouldMatchList, shouldNotBe,
                                       shouldReturn, shouldSatisfy, shouldThrow)
QuickCheck  Test.QuickCheck.Arbitrary (Arbitrary, arbitrary)
QuickCheck  Test.QuickCheck           (Positive(Positive), (==>), property)
```

#### Alternative Preludes

From the [alternative preludes](https://guide.aelve.com/haskell/alternative-preludes-zr69k1hc), it seems that [base-prelude](http://hackage.haskell.org/package/base-prelude) is the one that could allow use to get some immediate reduction in imports, but  that wouldn't help with the `hspec` package, that all the test suites import.

#### Custom Prelude

Writing a custom `Prelude` for our test suites could allow us to:

- Factor common patterns in the test suites.
- Reduce the number of imports.

This would leave us cleaner and easier to maintain test suites.

##### Why not?

The unavoidable problem is that we would need to deliver at least one additional file to each exercise, and this poses the same problem we have with the `stack.yaml` files.

#### Proof-of-concept

I just rewrote `pov` to use a custom prelude in #572.

##### Should we do it?

I'm not sure if the benefits are enough to justify the additional complexity, but at least it was a nice exercise to write it. 😄

**Comment by petertseng** [comment]
Seems difficult to manage since we didn't find symlinks to work for the https://github.com/exercism/problem-specifications/issues/601 . There could be some ways to make it work:

* It would be possible to write some tool to make sure that the additional file per exercise is the same (is it a goal to have it be the same?).
* Or perhaps if it could become a package that was included in Stackage, then everything will just work (but I'm not volunteering to maintain a package)
* If at https://github.com/exercism/haskell/pull/572#discussion_r125172729 the `stack.yaml` will allow `..` in relative path, in conjunction with the `exercism setup` command, the prelude could be placed there. This is a solution possible for this prelude but (AFAIK) not possible for stack.yaml because stack.yaml had to be in a specific location.

**Comment by petertseng** [comment]
I forgot to link for the issue for the exercism setup command, I am sorry.

https://github.com/exercism/cli/issues/388

**Comment by rbasso** [comment]
> ... (is it a goal to have it be the same?).

I suppose they should be exactly the same, otherwise we are just moving code around, without factoring anything.

> Or perhaps if it could become a package that was included in Stackage, then everything will just work (but I'm not volunteering to maintain a package)

I don't feel very comfortable about the idea of putting it in Hackage, because I think it is so specific that other people wouldn't be interested in it. But I guess we could put it online, in a repository.

> It would be possible to write some tool to make sure that the additional file per exercise is the same...

That would probably be the more direct solution, keeping each exercise self-contained.

> If at #572 (comment) the stack.yaml will allow .. in relative path...

I never tested, but I think it would work.

> ... in conjunction with the exercism setup command (must be implemented: exercism/cli#388), the prelude could be placed there. This is a solution possible for this prelude but (AFAIK) not possible for stack.yaml because stack.yaml had to be in a specific location.

Maybe we should focus on how to solve the `stack.yaml` problem, because it would solve both problems...

**Comment by rbasso** [comment]
Considering that there are technical problem in my proposal (rebuilds the custom prelude for each exercise) and that I don't have time to deal with that for a while, I'm closing this issue and the corresponding PR.
